tests/recovery/recovery_test/recovery_test.go

```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
+	snapID, err := bm.SetUpSystemWithOneSnapshot()
+	if err != nil {
+		t.FailNow()
+	}
+
+	cmpDir := bm.PathToTakeSnapshot
+
+	copyDir := t.TempDir()
+	if copyDir == "" {
+		t.FailNow()
+	}
+
+	err = bm.FileHandler.CopyAllFiles(cmpDir, copyDir)
+	require.NoError(t, err)
+
+	// add files
+	fileSize := 1 * 1024 * 1024
+	numFiles := 200
+
+	err = bm.GenerateRandomFiles(fileSize, numFiles)
+	require.NoError(t, err)
+
+	err = bm.FileHandler.CopyAllFiles(bm.PathToTakeSnapshot, copyDir)
+	require.NoError(t, err)
+
+	// modify original files
+	content := "\nthis is a test for TestConsistencyWhenKill9AfterModify\n"
+	err = bm.FileHandler.ModifyDataSetWithContent(copyDir, content)
+	require.NoError(t, err)
+	log.Println("Copy content and modify the files.")
+
+	// kill the kopia command before it exits
+	kopiaExe := os.Getenv("KOPIA_EXE")
+
+	cmd := exec.Command(kopiaExe, "repo", "connect", "filesystem", "--path="+dataRepoPath, "--content-cache-size-mb", "500", "--metadata-cache-size-mb", "500", "--no-check-for-updates")
+	env := []string{"KOPIA_PASSWORD=" + testenv.TestRepoPassword}
+	cmd.Env = append(os.Environ(), env...)
+
+	o, err := cmd.CombinedOutput()
+	require.NoError(t, err)
+	log.Println(string(o))
+
+	cmd = exec.Command(kopiaExe, "snap", "create", copyDir, "--json", "--parallel=1")
+
+	log.Println("Kill the kopia command before it exits:")
+	killOnCondition(t, cmd)
+
+	log.Println("Verify snapshot corruption:")
+	// verify snapshot corruption
+	err = bm.VerifySnapshot()
+	require.NoError(t, err)
+
+	// Create a temporary dir to restore a snapshot
+	restoreDir := t.TempDir()
+	if restoreDir == "" {
+		t.FailNow()
+	}
+
+	// try to restore a snapshot without any error messages.
+	stdout, err := bm.RestoreGivenOrRandomSnapshot(snapID, restoreDir)
+	require.NoError(t, err)
+
+	log.Println(stdout)
+
+	log.Println("Compare restored data and original data:")
+	require.NoError(t, bm.FileHandler.CompareDirs(restoreDir, cmpDir))
+}
+
+func killOnCondition(t *testing.T, cmd *exec.Cmd) {
+	t.Helper()
+
+	stderrPipe, err := cmd.StderrPipe()
+	require.NoError(t, err)
+
+	// execute kill -9 while receive ` | 1 hashing, 0 hashed (65.5 KB), 0 cached (0 B), uploaded 0 B, estimating...` message
+	var wg sync.WaitGroup
+
+	var mu sync.Mutex
+
+	// Add a WaitGroup counter for the first goroutine
+	wg.Add(1)
+
+	errOut := bytes.Buffer{}
+
+	go func() {
+		mu.Lock()
+		defer mu.Unlock()
+		defer wg.Done()
+
+		// Create a scanner to read from stderrPipe
+		scanner := bufio.NewScanner(stderrPipe)
+		for scanner.Scan() {
+			output := scanner.Text()
+			log.Println(output)
+			errOut.Write(scanner.Bytes())
+			errOut.WriteByte('\n')
+
+			log.Println(output)
+			// Check if the output contains the "hashing" etc.
+			if strings.Contains(output, "hashing") && strings.Contains(output, "hashed") && strings.Contains(output, "uploaded") || strings.Contains(output, "Snapshotting") {
+				log.Println("Detaching and terminating target process")
+				cmd.Process.Kill()
```

Comment:
This will work most of the times but will fail some of the times. The problem is that your strings.Contains() will not match every word every time this test runs.

`scanner := bufio.NewScanner(stderrPipe)` will allocate the default buffer size of 64k. What this means is that it MIGHT read upto those many characters before the first `scanner.Scan()` succeeds. In an unexpected scenario for your code, it might succeed once with half the word "uploa" first and the remaining "ded" in the next iteration, which will not make your test trigger the kill() action.

Which means that for your experiment, the kopia subcommand could actually print everything and then exit before the first iteration even triggering the kill action. Which is too late for you. Instead, I think what you are trying to do is to kill the subprocess as soon as you see lines being printed with the words you want.

So, you might be able to set a word split using `scanner.Split(bufio.ScanWords)` and check for words "uploaded" in that or you may also use `scanner.Split(bufio.ScanLines)` and then check for all the words in each line.